sdks/python/apache_beam/ml/transforms/base.py

```diff
@@ -254,3 +371,243 @@ def _increment_counters():
+            gcs_or_local_path=artifact_location,
+            file_name=_ATTRIBUTE_FILE_NAME,
+            stream=f,
+            mime_type='application/json')
+    else:
+      if not FileSystems.exists(artifact_location):
+        FileSystems.mkdirs(artifact_location)
+      # FileSystems.open() fails if the file does not exist.
+      with open(os.path.join(artifact_location, _ATTRIBUTE_FILE_NAME),
+                'w+') as f:
+        f.write(jsonpickle.encode(ptransform_list))
+
+  @staticmethod
+  def load_attributes(artifact_location):
+    with FileSystems.open(os.path.join(artifact_location, _ATTRIBUTE_FILE_NAME),
+                          'rb') as f:
+      return jsonpickle.decode(f.read())
+
+
+_transform_attribute_manager = _JsonPickleTransformAttributeManager
+
+
+class _MLTransformToPTransformMapper:
+  """
+  This class takes in a list of data processing transforms compatible to be
+  wrapped around MLTransform and returns a list of PTransforms that are used to
+  run the data processing transforms.
+
+  The _MLTransformToPTransformMapper is responsible for loading and saving the
+  PTransforms or attributes of PTransforms to the artifact location to seal
+  the gap between the training and inference pipelines.
+  """
+  def __init__(
+      self,
+      transforms: List[Union[BaseOperation, EmbeddingsManager]],
+      artifact_location: str,
+      artifact_mode: str,
+      pipeline_options: Optional[PipelineOptions] = None,
+  ):
+    self.transforms = transforms
+    self._parent_artifact_location = artifact_location
+    self.artifact_mode = artifact_mode
+    self.pipeline_options = pipeline_options
+
+  def create_and_save_ptransform_list(self):
+    ptransform_list = self.create_ptransform_list()
+    self.save_transforms_in_artifact_location(ptransform_list)
+    return ptransform_list
+
+  def create_ptransform_list(self):
+    previous_ptransform_type = None
+    current_ptransform = None
+    ptransform_list = []
+    for transform in self.transforms:
+      if not isinstance(transform, PTransformProvider):
+        raise RuntimeError(
+            'Transforms must be instances of PTransformProvider and '
+            'implement get_ptransform_for_processing() method.')
+      # for each instance of PTransform, create a new artifact location
+      current_ptransform = transform.get_ptransform_for_processing(
+          artifact_location=os.path.join(
+              self._parent_artifact_location, uuid.uuid4().hex[:6]),
+          artifact_mode=self.artifact_mode)
+      # Determine if a new ptransform should be added to the list
+      is_different_type = (type(current_ptransform) != previous_ptransform_type)
+      if is_different_type or not transform.requires_chaining():
+        ptransform_list.append(current_ptransform)
+        previous_ptransform_type = type(current_ptransform)
+
+      if hasattr(ptransform_list[-1], 'append_transform'):
+        ptransform_list[-1].append_transform(transform)
```

Comment:
Also, do actually need the `requires_chaining` function? Isn't it enough to just know that the transform itself has an `append_transform` function?

Reply:
One possible idea is to mandate that all `PTransformProvider` objects have an `append_transform` function that returns `True/False` depending on whether the transform could be appended and just return `False` if its a no-op (which could be done in the base class). Then we could (a) check if its the same type, (b) try appending, and (c) add it to the list if appending failed

Reply:
Removing the requires_chaining() method. Instead I am depending on if the ptransform contains append_transform method, if yes, I append the transform to the ptransform_list[-1]

Reply:
I'm still a little confused by the logic here. Assume for a given transform X `append_transform` is True, and `type(current_ptransform) != previous_ptransform_type)` is True. In that case we will execute the following:

```
ptransform_list.append(current_ptransform)
previous_ptransform_type = type(current_ptransform)
ptransform_list[-1].append_transform(transform)
```

This will:
1) Append the transform X to our list
2) Set `previous_ptransform_type = type(X)` (fine)
3) call X.append_transform(X)
```

why do we need to have the final `X.append_transform(X)`?

Reply:
A comment here would probably help